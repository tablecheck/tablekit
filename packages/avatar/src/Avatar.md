Avatars represent users that are using the application.

- If the image is passed, then the image will be shown
- If the initials are passed, then the initials will be shown
- If no image or initials are passed, then the user icon will be shown

### Examples

```tsx
import { Fragment, useState } from 'react';
import styled from '@emotion/styled';
import { ThemeProvider as EmotionThemeProvider } from '@emotion/react';
import { Header, Text } from '@tablecheck/tablekit-typography';
import {
  Size,
  DARK_COLORS,
  CLASSIC_COLORS,
  ThemeProvider
} from '@tablecheck/tablekit-theme';
import { Toggle } from '@tablecheck/tablekit-toggle';
import { ItemGroup } from '@tablecheck/tablekit-item';
import {
  avatarThemeNamespace,
  avatarDarkTheme,
  avatarClassicTheme
} from './index';

const darkTheme = {
  [avatarThemeNamespace]: avatarDarkTheme,
  colors: DARK_COLORS
};

const classicTheme = {
  colors: CLASSIC_COLORS,
  [avatarThemeNamespace]: avatarClassicTheme
};

const customLightTheme = {
  [avatarThemeNamespace]: {
    backgroundColor: CLASSIC_COLORS.success,
    textColor: 'white'
  }
};

const customDarkTheme = {
  [avatarThemeNamespace]: {
    backgroundColor: DARK_COLORS.warning,
    textColor: 'white'
  }
};

const ExamplesWrapper = styled.div`
  ${({ theme }) => `background-color: ${theme.colors.canvas};`}
  padding: 10px;
`;

const AvatarList = styled.ul`
  list-style-type: none;
  padding: 0;
  display: flex;
  align-items: center;
  margin-bottom: 10px;

  @media (max-width: 1080px) {
    display: block;
  }
`;

const AvatarWrapper = styled.li`
  margin-right: 10px;
  display: inline-block;

  @media (max-width: 1080px) {
    margin-bottom: 10px;
  }

  &:last-child {
    margin-right: 0;
  }
`;

const john = 'https://i.pravatar.cc/200?img=14';
const airi = 'https://i.pravatar.cc/200?img=31';

const avatarSizes = [
  Size.XSmall,
  Size.Small,
  Size.Regular,
  Size.Large,
  Size.XLarge,
  Size.XXLarge
];

const PresentationContainer = () => {
  const [isDarkTheme, toggleTheme] = useState(false);
  return (
    <EmotionThemeProvider theme={isDarkTheme ? darkTheme : classicTheme}>
      <ExamplesWrapper>
        <ItemGroup orientation="horizontal" hasIndent>
          <Toggle
            isChecked={isDarkTheme}
            onChange={() => toggleTheme(!isDarkTheme)}
          />
          <p style={{ display: 'flex', alignItems: 'center' }}>
            Use Dark Theme
          </p>
        </ItemGroup>
        <h3>Predefined Themes</h3>
        <AvatarList>
          {avatarSizes.map((size) => (
            <Fragment key={size}>
              <AvatarWrapper>
                <Avatar size={size} name="John Smith" initials="JS" />
              </AvatarWrapper>
              {size !== 'xxlarge' && (
                <AvatarWrapper>
                  <Avatar
                    size={size}
                    name="John Smith"
                    initials="JS"
                    imageUrl={john}
                  />
                </AvatarWrapper>
              )}
            </Fragment>
          ))}
        </AvatarList>
        <h3>Custom Themes</h3>
        <EmotionThemeProvider
          theme={isDarkTheme ? customDarkTheme : customLightTheme}
        >
          <AvatarList>
            {avatarSizes.map((size) => (
              <Fragment key={size}>
                <AvatarWrapper>
                  <Avatar size={size} name="愛梨" initials="愛梨" />
                </AvatarWrapper>
                {size !== 'xxlarge' && (
                  <AvatarWrapper>
                    <Avatar
                      size={size}
                      name="愛梨"
                      initials="愛梨"
                      imageUrl={airi}
                    />
                  </AvatarWrapper>
                )}
              </Fragment>
            ))}
          </AvatarList>
          <h3>How do display the user icon with no initials or image</h3>
          <AvatarList>
            {avatarSizes.map((size) => (
              <Fragment key={size}>
                <AvatarWrapper>
                  <Avatar size={size} />
                </AvatarWrapper>
                {size !== 'xxlarge' && (
                  <AvatarWrapper>
                    <Avatar size={size} imageUrl={airi} />
                  </AvatarWrapper>
                )}
              </Fragment>
            ))}
          </AvatarList>
        </EmotionThemeProvider>
      </ExamplesWrapper>
    </EmotionThemeProvider>
  );
};

<ThemeProvider useLocalStyles>
  <PresentationContainer />
</ThemeProvider>;
```
