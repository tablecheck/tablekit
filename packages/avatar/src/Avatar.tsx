import { Icon, getIcon } from '@tablecheck/tablekit-icon';
import { Size } from '@tablecheck/tablekit-theme';

import { Container, Image } from './styled';
import { AvatarProps } from './types';

export const Avatar = ({
  size,
  imageUrl,
  initials,
  name,
  ...restProps
}: AvatarProps): JSX.Element => {
  let formattedInitials = initials || '';
  formattedInitials =
    size === Size.XSmall || size === Size.Small || size === Size.Regular
      ? formattedInitials.substring(0, 1).toUpperCase()
      : formattedInitials.toUpperCase();

  let content: JSX.Element | string = <Icon icon={getIcon('user')} />;

  if (initials && size !== Size.XSmall) {
    content = formattedInitials;
  }

  if (imageUrl) {
    content = <Image src={imageUrl} alt={name} title={name} />;
  }

  return (
    <Container {...restProps} size={size || Size.Regular} imageUrl={imageUrl}>
      {content}
    </Container>
  );
};
