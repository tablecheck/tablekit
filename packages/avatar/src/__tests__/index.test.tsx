import { mount } from '@tablecheck/tablekit-enzyme';
import '@tablecheck/tablekit-free-icon-config';

import { Avatar } from '../Avatar';

beforeEach(() => {
  global.console.error = jest.fn();
});

describe('Avatar', () => {
  test('should be possible to create a component', () => {
    const wrapper = mount(<Avatar />);
    expect(wrapper).not.toBeUndefined();
  });

  test('should correctly apply styling', () => {
    expect(mount(<Avatar />)).toMatchSnapshot();
  });

  test('should allow passing of custom className', () => {
    expect(mount(<Avatar className="test-class" />)).toMatchSnapshot();
  });
});
