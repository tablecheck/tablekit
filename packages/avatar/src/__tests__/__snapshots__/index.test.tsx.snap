// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Avatar should allow passing of custom className 1`] = `
.emotion-0 {
  width: 32px;
  height: 32px;
  font-size: 14px;
  background-color: #5C11A6;
  border-radius: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  color: white;
  overflow: hidden;
}

.emotion-0 svg {
  width: 14px;
  height: 14px;
}

<Avatar
  className="test-class"
>
  <Container
    className="test-class"
    size="regular"
  >
    <div
      className="test-class emotion-0 emotion-1"
      size="regular"
    >
      <Icon
        icon="fas-user"
      >
        icon-user
      </Icon>
    </div>
  </Container>
</Avatar>
`;

exports[`Avatar should correctly apply styling 1`] = `
.emotion-0 {
  width: 32px;
  height: 32px;
  font-size: 14px;
  background-color: #5C11A6;
  border-radius: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  color: white;
  overflow: hidden;
}

.emotion-0 svg {
  width: 14px;
  height: 14px;
}

<Avatar>
  <Container
    size="regular"
  >
    <div
      className="emotion-0 emotion-1"
      size="regular"
    >
      <Icon
        icon="fas-user"
      >
        icon-user
      </Icon>
    </div>
  </Container>
</Avatar>
`;
